idea-plugin/src/main/kotlin/arrow/meta/ide/internal/registry/IdeInternalRegistry.kt

```diff
@@ -50,19 +55,33 @@ internal interface IdeInternalRegistry : InternalRegistry {
       )
     } ?: ExtensionPhase.Empty
 
-  override fun CompilerContext.registerIdeExclusivePhase(currentPhase: ExtensionPhase): Unit =
-    when (currentPhase) {
-      is ExtensionPhase.Empty, is CollectAdditionalSources, is Composite, is Config, is ExtraImports,
-      is PreprocessedVirtualFileFactory, is StorageComponentContainer, is AnalysisHandler, is ClassBuilder,
-      is Codegen, is DeclarationAttributeAlterer, is PackageProvider, is SyntheticResolver,
-      is IRGeneration, is SyntheticScopeProvider -> Unit // filter out ExtensionPhases which happen in the compiler
-      is ExtensionProvider<*> -> registerExtensionProvider(currentPhase)
-      is AnActionExtensionProvider -> registerAnActionExtensionProvider(currentPhase)
-      is IntentionExtensionProvider -> registerIntentionExtensionProvider(currentPhase)
-      is SyntaxHighlighterExtensionProvider -> registerSyntaxHighlighterExtensionProvider(currentPhase)
-      else -> LOG.error("Unsupported ide extension phase: $currentPhase")
+  fun registerMetaApplication(application: Application, compilerConfiguration: CompilerConfiguration): Unit =
+    intercept().forEach { plugin ->
+      println("Registering plugin: $plugin extensions: ${plugin.meta}")
+      plugin.meta.invoke().forEach(::registerIdeExclusiveApplicationPhase)
     }
 
+  fun registerIdeExclusiveApplicationPhase(currentPhase: ExtensionPhase): Unit = when (currentPhase) {
```

Comment:
There is also the option to create one unified `registerIdeExclusivePhase` we're we add State explicitly in the FunctionParameter. Though this poor design makes only sense if we want to keep the editor open for ProjectExtensions, even though their fairly used in the ide.